### PR TITLE
first pass non-square map for flat sky

### PR DESCRIPTION
**Pull request by kimmywu**
This PR addresses issue #15. In this first pass, I have run all the cases in `CMBLensing.jl/docs/src/01-03` and the code produced sensible results when Nside=(Nx,Ny). 

New: 
- Nside can now be a Number or a Tuple. 
-- If Nside <: Number, then the previous square-map behavior is preserved.
-- If Nside <: Tuple, it expects Nside = (Nx, Ny).

Next steps:
- Add documentation about Nside
- Adjust the aspect ratio of figsize when plotting non-square map.  
- Test GPU cases

@marius311: There are definitely places where I think someone who is more familiar with Julia/CMBLensing.jl can streamline. There are also places I think should work but probably didn't get tested when running the examples above.

### CODE REVIEW

File: test/runtests.jl [hunk 1]
```diff
@@ -523,42 +526,46 @@ end
...
-            f° = L(ϕ)*D*f
-
-            @test lnP(0,f,ϕ,ds) ≈ lnP(1,    f̃,  ϕ ,ds) rtol=1e-4
-            @test lnP(0,f,ϕ,ds) ≈ lnP(:mix, f°, ϕ, ds) rtol=1e-4
-
-            ε = sqrt(eps(T))
-            seed!(0)
-            δf,δϕ = simulate(Cf),simulate(Cϕ)
+
+    for nside = [128, (128, 64), (64, 128)]
```
Comment (marius311) [inline test/runtests.jl:530]: For all the new `for` loops, can you add a testset so it prints out on which `nside` it failed (when/if it does?). I think the syntax is something like

```julia
@testset "Nside $nside" for nside = [128, (128, 64), (64, 128)]
```

(You may want to double check / see https://docs.julialang.org/en/v1/stdlib/Test/ I haven't tried running that)

File: src/flat_generic.jl [hunk 2]
```diff
@@ -5,7 +5,7 @@ const FlatFieldFourier{P,T,M} = Union{FlatFourier{P,T,M},FlatS2{P,T,M},FlatS02Fo
 
 ### pretty printing
 @show_datatype show_datatype(io::IO, t::Type{F}) where {N,θ,∂mode,D,T,M,F<:FlatField{Flat{N,θ,∂mode,D},T,M}} =
-    print(io, "$(pretty_type_name(F)){$(N)×$(N)$(D==1 ? "" : "(×$D)") map, $(θ)′ pixels, $(∂mode.name.name), $M}")
+print(io, "$(pretty_type_name(F)){$( typeof(N) <: Tuple ? N[2] : N )×$(  typeof(N) <: Tuple ? N[1] : N )$(D==1 ? "" : "(×$D)") map, $(θ)′ pixels, $(∂mode.name.name), $M}")
```
Comment (marius311) [inline src/flat_generic.jl:8]: Can you swap this to the shorter `join(Nside .* (1,1), "×")` ?

File: src/masking.jl [hunk 3]
```diff
@@ -5,7 +5,7 @@ function make_mask(
     edge_rounding_deg = 1,
     apodization_deg = 1,
     ptsrc_radius_arcmin = 7, # roughly similar to Story et al. 2015
-    num_ptsrcs = round(Int,((Nside*θpix)/60)^2 * 120/100) # SPT-like density of sources
+    num_ptsrcs = typeof(Nside)<:Tuple ? round(Int, Nside[1]*Nside[2]*(θpix/60)^2 * 120/100)  : round(Int,((Nside*θpix)/60)^2 * 120/100) # SPT-like density of sources
```
Comment (marius311) [inline src/masking.jl:8]: I think here you can use `prod(Nside .* (1,1))`.

Comment (marius311) [inline src/masking.jl:8]: I think here you can use `prod(Nside .* (1,1))`.

File: src/masking.jl [hunk 4]
```diff
@@ -28,7 +28,7 @@ make_mask(::FlatField{<:Flat{Nside,θpix}}; kwargs...) where {Nside,θpix} = mak
 # all padding/smoothing/etc... quantities below are in units of numbers of pixels
 
 function boundarymask(Nside, pad)
-    m = fill(true,Nside,Nside)
+    m = typeof(Nside)<:Tuple ? fill(true,Nside[2],Nside[1]) : fill(true, Nside, Nside)
```
Comment (marius311) [inline src/masking.jl:31]: And here maybe `fill(true, (Nside .* (1,1))...)`

(Sorry this is so minor, just trying to be more concise and get rid of the tiny code duplication. If you do want to keep these anywhere, note `Nside isa Tuple` is equivalent to `typeof(Nside)<:Tuple` and slighly more common)

**Comment by marius311** [comment]
Left a couple of minor comments, otherwise looks good to go. The test failures are unrelated I think, I still haven't figured out how to fix the seeds in the random fields generated in the posterior tests, so they just randomly fail sometimes. We can merge this and I can figure that out later.

**Comment by marius311** [comment]
Hmm one last thing but the inconsistency below below between setting `Nside=(128,64)` vs. `f`, `f[:Ix]`, and the plot showing `64×128` is kind of bugging me. I feel like whatever these are, they should be the same thing.

image

I see two options:
1. We could switch from storing `Nside = (Nx, Ny)` to storing `Nside = (Ny, Nx)`, in which case everything above would be consistent, but its maybe surprising that something printed 64×128 is a wide short map, my intuition would be that it would be X×Y (maybe that's just me?) 
2. We could switch the field pretty printing and plot titles to show Nx×Ny, but then `f` and `f[:Ix]` print different things, and 2D indexing into the field object is weird since `f[100,50]` is an out of bonuds index for a map printed as 128×64. 

I kind of think 1 is the least bad option, then you just have to remember "absolutely everything in CMBLensing.jl is Nrows×Ncols", and we just live with that. Thoughts? Maybe @cailmdaley too?

File: test/runtests.jl (hunk 1 above)
Comment (kimmywu) [reply test/runtests.jl]: Yup. Done.

File: src/masking.jl (hunk 4 above)
Comment (kimmywu) [reply src/masking.jl:31]: No, these are all good suggestions!

**Comment by kimmywu** [comment]
> 1. We could switch from storing `Nside = (Nx, Ny)` to storing `Nside = (Ny, Nx)`, in which case everything above would be consistent, but its maybe surprising that something printed 64×128 is a wide short map, my intuition would be that it would be X×Y (maybe that's just me?)
> 2. We could switch the field pretty printing and plot titles to show Nx×Ny, but then `f` and `f[:Ix]` print different things, and 2D indexing into the field object is weird since `f[100,50]` is an out of bonuds index for a map printed as 128×64.
> 
> I kind of think 1 is the least bad option, then you just have to remember "absolutely everything in CMBLensing.jl is Nrows×Ncols", and we just live with that. Thoughts? Maybe @cailmdaley too?

I'll go with 1, which I think is the cleanest. And we just need to be clear about how Nside is (Ny, Nx). 
(But yes, my intuition is also that it would be X×Y.)

**Comment by cailmdaley** [comment]
This looks great! I agree that option 1 seems most reasonable. Maybe this is blasphemy, but I do tend to think of maps as matrices..

**Comment by cailmdaley** [comment]
Fun thing I just ran into: `Npix = prod(Nside)` works regardless of whether the map is square or not. So nice

**Comment by kimmywu** [comment]
Note I didn't check the GPU option in the bilinearlens fix commit (27493af), because cori...

**Comment by marius311** [comment]
Thanks a bunch again for this @kimmywu !

**Pull request merged**